File: app/src/main/java/com/dougkeen/bart/backend/TransitProjection.java
package com.dougkeen.bart.backend;

import java.util.Objects;

/** Derives a consumer-facing value from one complete feed snapshot. */
public interface TransitProjection<T> {
    T project(TransitFeedSnapshot snapshot) throws Exception;

    /** Allows projections to suppress updates caused by unrelated feed data. */
    default boolean areEquivalent(T previous, T current) {
        return Objects.equals(previous, current);
    }
}

File: app/src/main/kotlin/com/dougkeen/bart/backend/TransitProjection.kt
package com.dougkeen.bart.backend

import java.util.Objects
import kotlin.jvm.JvmDefaultWithCompatibility

/** Derives a consumer-facing value from one complete feed snapshot. */
@JvmDefaultWithCompatibility
interface TransitProjection<T> {
    @Throws(Exception::class)
    fun project(snapshot: TransitFeedSnapshot): T

    /** Allows projections to suppress updates caused by unrelated feed data. */
    fun areEquivalent(previous: T?, current: T?): Boolean =
        Objects.equals(previous, current)
}
